src/com/intellij/rt/coverage/data/ClassData.java

```diff
@@ -306,13 +306,66 @@ public void setSource(String source) {
   public String getSource() {
     return mySource;
   }
-    
-  public int[] touchLines(int[] lines) { //todo
-      myLineMask = lines;
-      return lines;
+
+  public synchronized void createHitsMask(int size) {
```

Comment:
when this createHitsMask can be called? You sync it once here but not `ClassData#touchLine` and volatile doesn't guarantee anything about access to the array elements. Or what is the logic?

Reply:
This method is called in the end of new tracing class transformation. ClassData#touchLine is not called in new tracing mode, so createHitsMask and getLineMask are the only methods that access this field in application runtime(before coverage collection). Synchronized is added to prevent errors in case of class redefinition, getLineMask is not synchronized as the field is volatile, so the array's size is OK for this thread.

Reply:
extremely error prone, I'd leave a comment on threading assumptions or better use different fields to avoid confusion

Reply:
Okay, I'll use a separate array for new tracing then.